superset-frontend/src/SqlLab/components/SaveDatasetModal.tsx

```diff
@@ -0,0 +1,183 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React, { useState, FunctionComponent } from 'react';
+import { Radio, AutoComplete, Input } from 'src/common/components';
+import StyledModal from 'src/common/components/Modal';
+import Button from 'src/components/Button';
+import { styled } from '@superset-ui/core';
+import { RadioChangeEvent } from 'antd/lib/radio';
+
+interface SaveDatasetModalProps {
+  visible: boolean;
+  onOk: () => void;
+  onHide: () => void;
+  handleDatasetNameChange: (e: { target: { value: any } }) => void;
+  userDatasetsOwned: Array<Record<string, any>>;
+  handleSaveDatasetRadioBtnState: (e: RadioChangeEvent) => void;
+  saveDatasetRadioBtnState: number;
+  overwriteDataSet: boolean;
+  handleOverwriteCancel: () => void;
+  handleOverwriteDataset: () => void;
+  handleOverwriteDatasetOption: (
+    data: string,
+    option: Record<string, any>,
+  ) => void;
+  defaultCreateDatasetValue: string;
+}
+
+const Styles = styled.div`
+  .smd-input {
+    margin-left: 45px;
+    width: 290px;
+  }
+  .smd-autocomplete {
+    margin-left: 8px;
+    width: 290px;
+  }
+  .smd-radio {
+    display: block;
+    height: 30px;
+    margin: 10px 0px;
+    line-height: 30px;
+  }
+`;
+
+// eslint-disable-next-line no-empty-pattern
+export const SaveDatasetModal: FunctionComponent<SaveDatasetModalProps> = ({
+  visible,
+  onOk,
+  onHide,
+  handleDatasetNameChange,
+  userDatasetsOwned,
+  handleSaveDatasetRadioBtnState,
+  saveDatasetRadioBtnState,
+  overwriteDataSet,
+  handleOverwriteCancel,
+  handleOverwriteDataset,
+  handleOverwriteDatasetOption,
+  defaultCreateDatasetValue,
+}) => {
+  const [options, setOptions] = useState<
+    {
+      value: string;
+      dataSetId: number;
+    }[]
+  >([]);
+
+  const onSearch = (searchText: any) => {
+    setOptions(
+      !searchText
+        ? []
+        : userDatasetsOwned.map(d => ({
+            value: d.dataSetName,
+            dataSetId: d.dataSetId,
+          })),
+    );
+  };
+
+  const filterAutocompleteOption = (
+    inputValue: string,
+    option: { [key: string]: any } | any,
```

Comment:
do you want `| any` or `| undefined | null`?

Reply:
`{ [key: string]: any } | any` is the same as `any`, no?

Do you need the `any` here?

Reply:
mostly a hack to pass typing, but still let the user/dev know what is expected

Reply:
yeah, exactly, `any` basically undoes the `{ [key: string]: any }` validation. If you want to allow someone to pass an array, then cool, but if you want to allow null or undefined, that would be better to be explicit that those are the types that you expect.

Reply:
But you shouldn't need to do this to pass typing — if it's not passing when you set the type to `{ [key: string]: any }` it means there's some wrong assumption elsewhere in your code.

Maybe defining the type as an interface would help? Where are you getting typing errors?